README
What *is* hunt?

	Hunt is a multi-player search-and-destroy game that takes place
	in a maze.  The game may either be slow and strategic or fast
	and tactical, depending on how familiar the players are with the
	keyboard commands.

Distribution Policy:

	Hunt is part of the user-contributed software distributed by
	Berkeley in 4BSD.  The sources are copyrighted by the authors
	and the University of California.  You may redistribute freely
	as long as the copyright notices are retained.

Words of Warning:

	hunt uses the socket mechanism of 4BSD Unix, so if you are on
		System V (my sympathies), you're on your own.
	If your machine does not permit non-setuid-root processes to
		broadcast UDP packets, then hunt uses a *very* inefficient
		method for locating the hunt server: it sends a packet
		to every host on your network.  If your machine falls
		into this category, we strongly recommend that you use
		either standalone or inetd mode *and* start hunt by
		specifying the hunt server host.
	hunt can be configured to use Unix-domain sockets, but that
		code has not been tested in recent memory.  Also, since
		4.2BSD Unix-domain sockets are buggy, running hunt on
		4.2BSD with Unix-domain sockets will probably crash
		your system.  If you want to experiment, feel free to
		do so.  However, don't say I didn't warn you :-).
	hunt uses a fair amount of CPU time, both in user time (for
		computing interactions) and system time (for processing
		terminal interrupts).  We found that a VAX 750 can
		support about three users before the system is
		noticeably impacted.  The number goes up to about 8 or
		10 for a VAX 8650.  On a network of Sun 3/50's with the
		server running on a 3/280, things work much more
		smoothly as the computing load is distributed across
		many machines.
	hunt may be dangerous to your health.  "Arthritic pain" and
		"lack of circulation" in fingers have been reported by
		hunt abusers.  Hunt may also be addictive, and the
		withdrawal symptoms are not pretty :-)

Installation:
	[ Standard './configure && make && make install' procedure should now
	work - see INSTALL for details. Original section included below for
	historical interest!  -- MDB 2010 ]

	1. Edit file "Makefile" and make sure the options selected are
		reasonable.  There are four "make" variables that you
		should check: GAME_PARAM, SYSCFLAGS, SYSLDFLAGS, and DEFS.
		GAME_PARAM controls what features of the game will be
		compiled in (e.g. reflecting walls).  The optional features
		are listed in comments above where GAME_PARAM is defined.
		If you want to try them, just add the ones you want to the 
		GAME_PARAM definition.

		DEFS is where most system configuration is described.
		If your system is 4.3BSD, Sun, Ultrix, Convex, HPUX
		v6.0.1, or SGI, you're in luck.  We provide the
		appropriate definitions for these systems and you just
		need to select one of them (e.g. if you have an Ultrix
		system, just change the line
			DEFS=	$(GAME_PARAM) $(DEFS_43)
		to
			DEFS=	$(GAME_PARAM) $(DEFS_ULTRIX)
		).  If your system is *not* listed above, then you may
		need to do some experiments.  All of the options are
		documented in the Makefile, be brave.

		SYSCFLAGS and SYSLDFLAGS are used for "unusual" systems
		and you probably won't need to deal with it.  An
		example of an unusual system is the Silicon Graphics
		IRIS, which keeps the network socket code in a BSD
		emulation library that is in -lbsd.  Edit these only if
		you *know* your system is "different."

	2. Edit file "Makefile" and look at the "install:" target.  By
		default, files are installed in /usr/games,
		/usr/games/lib, and /usr/man/man6, which are "standard"
		locations for games.  If your system has a local games
		directory, you'll need to change these.
	3. Edit file "pathname.c" and make sure the file names and port
		numbers are reasonable.  You can ignore the first set
		of variables as they are used only for debugging
		purposes.  The second set is used in the installed
		version of hunt.  The important variables are "Driver"
		(where the server is kept), "Test_port" (the Internet
		UDP port number that new players should use to contact
		the server), and "Stat_file" (where scoring statistics
		and body counts are written).  The only tricky variable
		here is "Test_port".  The default value is chosen so
		that it is unlikely to conflict with other service port
		numbers, but you can change it if you want to.
	4. Type "make install", which will compile and install the
		programs and manual pages.  Now you're almost ready to
		go (see next section).  There may be some warnings during
		compilation.  Ignore them.

Setting up the network:

	Hunt may be set up in one of three modes: standalone, inetd, or
	nothing.  In "standalone" mode, there is always a hunt server
	running on a server machine.  All players who enter the game
	will be talking to this server.  This is the mode we use at
	UCSF.  The cost is one entry in the process table on the server
	machine.  In "inetd" mode, the server is started via inetd.
	Again, only one machine should be set up to answer game
	requests.  The cost is having to edit a few system files.  In
	"nothing" mode, no server is running when there is no one
	playing.  The first person to enter hunt will automatically
	start up a server on his machine.  This, of course, gives him
	an unfair advantage.  Also, there may be race conditions such
	that players end up in different games.  The choice of which
	mode to use depends on site configuration and politics.  We
	recommend using "standalone" mode because it is simple to set
	up and starts up rapidly.

	-----

	FOR STANDALONE MODE, put these lines in /etc/rc.local on the
	server machine.  THERE SHOULD ONLY BE ONE SERVER MACHINE!

	# start up the hunt daemon if present
	if [ -f /usr/games/lib/huntd ]; then
		/usr/games/lib/huntd -s & (echo -n ' huntd')	>/dev/console
	fi

	Also, you should start one up (on the off chance that you will
	want to test this mess :-) by typing "/usr/games/lib/hunt -s".

	-----

	FOR INETD MODE, then things get more complicated.  You need to
	edit both /etc/services and /etc/inetd.conf.  In /etc/services,
	add the line

	hunt		26740/udp

	26740 corresponds to the default "Test_port".  If you changed
	that variable, then you should put whatever value you used here
	as well.  In /etc/inetd.conf, add the line

	hunt	dgram	udp	wait	nobody	/usr/games/lib/huntd	huntd

	This works for 4.3BSD.  I don't remember the configuration file
	format for 4.2BSD inetd.

	See the huntd.6 manual page for more details.

	-----

	FOR NOTHING MODE, do nothing.

Testing:
	Now you are ready to test the code.  Type "/usr/games/hunt" or
	whatever you call the hunt executable.  You should be prompted
	for your name and team.  Then you should get the display of a
	maze.  At this point, you should read the manual page :-).

======

Hunt is not officially supported by anyone anywhere (that I know of);
however, bug reports will be read and bug fixes/enhancements may be
sent out at irregular intervals.  Send no flames, just money.  Happy
hunting.

					Conrad Huang
					conrad@cgl.ucsf.edu
					Greg Couch
					gregc@cgl.ucsf.edu
					October 17, 1988

P.S.  The authors of the game want to emphasize that this version of hunt
was started over eight years ago, and the programming style exhibited here
in no way reflects the current programming practices of the authors.


ChangeLog
Sat Apr 23 13:24:24 BST 2005  Martin Bays <mbays@freeshell.org>
  * more info on -O2 bug

    M ./BUGS +4

Sat Apr 23 13:12:50 BST 2005  Martin Bays <mbays@freeshell.org>
  tagged 1.1.1


Sat Apr 23 13:12:34 BST 2005  Martin Bays <mbays@freeshell.org>
  * version 1.1.1

    M ./configure.ac -1 +1

Sat Apr 23 13:11:39 BST 2005  Martin Bays <mbays@freeshell.org>
  * <bugnote> sendto warning

    M ./BUGS +8

Sat Apr 23 12:33:51 BST 2005  Martin Bays <mbays@freeshell.org>
  * make make clean work

    M ./huntd/Makefile.linux -1 +1

Sat Apr 23 12:33:16 BST 2005  Martin Bays <mbays@freeshell.org>
  * prettify

    M ./BUGS +3

Sat Apr 23 12:32:39 BST 2005  Martin Bays <mbays@freeshell.org>
  * workaround segfault with gcc -O2

    M ./BUGS +4
    M ./configure.ac +3
    M ./huntd/Makefile.linux -1 +1

Wed Apr 20 21:08:58 BST 2005  Martin Bays <mbays@freeshell.org>
  tagged 1.1


Wed Apr 20 20:57:24 BST 2005  Martin Bays <mbays@freeshell.org>
  * Don't need to add @LIBS@ manually

    M ./huntd/Makefile.am -1 +1

Wed Apr 20 20:35:51 BST 2005  Martin Bays <mbays@freeshell.org>
  UNDO: Failed attempt to use AC_REPLACE_FUNCS and LIBOBJS
  Failed because I can't work out what should go in lib/Makefile.am
  
  It appears to be impossible (for me) to find this kind of information in all
  the documentation on auto*
  
  Hnnnnnnnnnnnnnng.

    M ./configure.ac -8 +4
    M ./lib/Makefile.am -1 +1
    M ./lib/fgetln.c +4
    M ./lib/strlcat.c +4
    M ./lib/strlcpy.c +4

Wed Apr 20 20:35:51 BST 2005  Martin Bays <mbays@freeshell.org>
  * Failed attempt to use AC_REPLACE_FUNCS and LIBOBJS
  Failed because I can't work out what should go in lib/Makefile.am
  
  It appears to be impossible (for me) to find this kind of information in all
  the documentation on auto*
  
  Hnnnnnnnnnnnnnng.

    M ./configure.ac -4 +8
    M ./lib/Makefile.am -1 +1
    M ./lib/fgetln.c -4
    M ./lib/strlcat.c -4
    M ./lib/strlcpy.c -4

Wed Apr 20 19:53:08 BST 2005  Martin Bays <mbays@freeshell.org>
  * don't install libhunt

    M ./lib/Makefile.am -1 +1

Wed Apr 20 19:52:54 BST 2005  Martin Bays <mbays@freeshell.org>
  * install huntd-hacked in sbin

    M ./huntd/Makefile.am -1 +1

Wed Apr 20 17:02:45 BST 2005  Martin Bays <mbays@freeshell.org>
  * done - use autoconf

    M ./TODO -2

Wed Apr 20 16:59:03 BST 2005  Martin Bays <mbays@freeshell.org>
  * asm/socket.h appears to be superfluous - don't include it

    M ./configure.ac -1 +4
    M ./hunt/hunt.c -4
    M ./hunt/list.c -4

Wed Apr 20 16:45:39 BST 2005  Martin Bays <mbays@freeshell.org>
  * rearrange tests and give errors on failure

    M ./configure.ac -6 +10

Tue Apr 19 22:07:41 BST 2005  Martin Bays <mbays@freeshell.org>
  * <bugfix> call nanosleep right

    M ./hunt/otto.c -1 +1

Tue Apr 19 22:07:06 BST 2005  Martin Bays <mbays@freeshell.org>
  * use huntd/hunt.h not hunt/hunt.h

    M ./hunt/Makefile.am -1 +1

Tue Apr 19 21:41:21 BST 2005  Martin Bays <mbays@freeshell.org>
  * remove silly file

    M ./huntd/Makefile.am -1 +1
    R ./huntd/test

Tue Apr 19 21:35:40 BST 2005  Martin Bays <mbays@freeshell.org>
  * export only defined functions

    M ./lib/lib.h -1 +5

Tue Apr 19 21:35:09 BST 2005  Martin Bays <mbays@freeshell.org>
  * stray blash

    M ./acinclude.m4 -1 +1

Tue Apr 19 21:28:00 BST 2005  Martin Bays <mbays@freeshell.org>
  * <bugfix>

    M ./huntd/server.h -3

Tue Apr 19 21:27:37 BST 2005  Martin Bays <mbays@freeshell.org>
  * link to wrap only if we have it!

    M ./huntd/Makefile.am -1 +1

Tue Apr 19 21:20:38 BST 2005  Martin Bays <mbays@freeshell.org>
  * autoconf and automake scripts

    A ./Makefile.am
    A ./acinclude.m4
    A ./configure.ac
    A ./hunt/Makefile.am
    A ./huntd/Makefile.am
    A ./lib/Makefile.am

Tue Apr 19 21:20:02 BST 2005  Martin Bays <mbays@freeshell.org>
  * define utility functions only if not already defined

    M ./lib/fgetln.c -1 +1
    M ./lib/strlcat.c +4
    M ./lib/strlcpy.c +4

Tue Apr 19 21:13:13 BST 2005  Martin Bays <mbays@freeshell.org>
  * <bugfix> colour unless old curses

    M ./hunt/hunt.c -1 +1

Tue Apr 19 21:02:57 BST 2005  Martin Bays <mbays@freeshell.org>
  * remove unwanted files

    R ./huntd/test.c
    R ./lib/Makefile

Tue Apr 19 21:02:35 BST 2005  Martin Bays <mbays@freeshell.org>
  * HAVE_TCPD -> HAVE_LIBWRAP

    M ./huntd/answer.c r1
    M ./huntd/driver.c r1
    M ./huntd/server.h r1

Tue Apr 19 20:53:20 BST 2005  Martin Bays <mbays@freeshell.org>
  * new curses checks

    M ./hunt/display.c -1 +11

Tue Apr 19 20:52:01 BST 2005  Martin Bays <mbays@freeshell.org>
  * now using autoconf to check sockaddr_in struct

    M ./hunt/list.h -6

Tue Apr 19 20:51:46 BST 2005  Martin Bays <mbays@freeshell.org>
  * check directly for SO_USELOOPBACK

    M ./hunt/hunt.c -1 +1

Tue Apr 19 20:51:09 BST 2005  Martin Bays <mbays@freeshell.org>
  * new sockio / asm/socket checking

    M ./hunt/hunt.c -2 +4
    M ./hunt/list.c -2 +4

Tue Apr 19 20:48:34 BST 2005  Martin Bays <mbays@freeshell.org>
  * make every .c include config.h

    M ./hunt/connect.c +2
    M ./hunt/display.c -3 +1
    M ./hunt/hunt.c +2
    M ./hunt/list.c +2
    M ./hunt/otto.c +2
    M ./hunt/playit.c +2
    M ./huntd/answer.c +2
    M ./huntd/conf.c +2
    M ./huntd/draw.c +2
    M ./huntd/driver.c +2
    M ./huntd/execute.c +2
    M ./huntd/expl.c +2
    M ./huntd/extern.c +2
    M ./huntd/makemaze.c +2
    M ./huntd/message.c +2
    M ./huntd/shots.c +2
    M ./huntd/terminal.c +2
    M ./lib/fgetln.c +2
    M ./lib/strlcat.c +2
    M ./lib/strlcpy.c +2

Tue Apr 19 20:47:57 BST 2005  Martin Bays <mbays@freeshell.org>
  * remove silly configure script

    R ./configure

Tue Apr 19 20:46:47 BST 2005  Martin Bays <mbays@freeshell.org>
  tagged 1.0


Tue Apr 19 19:33:26 BST 2005  Martin Bays <mbays@freeshell.org>
  * always include sys/types.h - I am informed this is portable

    M ./hunt/connect.c -2

Tue Apr 19 18:06:04 BST 2005  Martin Bays <mbays@freeshell.org>

hunt/hunt.c
/*	$OpenBSD: hunt.c,v 1.11 2003/06/11 08:45:25 pjanzen Exp $	*/
/*	$NetBSD: hunt.c,v 1.8 1998/09/13 15:27:28 hubertf Exp $	*/
/*
 * Copyright (c) 1983-2003, Regents of the University of California.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are 
 * met:
 * 
 * + Redistributions of source code must retain the above copyright 
 *   notice, this list of conditions and the following disclaimer.
 * + Redistributions in binary form must reproduce the above copyright 
 *   notice, this list of conditions and the following disclaimer in the 
 *   documentation and/or other materials provided with the distribution.
 * + Neither the name of the University of California, San Francisco nor 
 *   the names of its contributors may be used to endorse or promote 
 *   products derived from this software without specific prior written 
 *   permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS 
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED 
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A 
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT 
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, 
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY 
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "../config.h"

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <curses.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <netdb.h>

#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>

#ifdef HAVE_SYS_SOCKIO_H
#include <sys/sockio.h>
#endif

#include <netinet/in.h>
#include <net/if.h>

#include <arpa/inet.h>

#include "hunt.h"
#include "display.h"
#include "client.h"
#include "list.h"
#include "lib.h"

#ifndef __GNUC__
#define __attribute__(x)
#endif

FLAG	Am_monitor = FALSE;
int	Socket;
char	map_key[256];			/* what to map keys to */
FLAG	no_beep = FALSE;
char	*Send_message = NULL;
char	team = '-';
#ifndef USE_OLD_CURSES
FLAG	Use_Color = TRUE;
#else
FLAG	Use_Color = FALSE;
#endif

static char	*Sock_host;
static char	*use_port;
static FLAG	Query_driver = FALSE;
static FLAG	Show_scores = FALSE;
static struct sockaddr	Daemon;
static socklen_t Daemon_sa_len;		/* not all systems have an sa_len
					   field to the sockaddr struct, so
					   use this instead */


static char	name[NAMELEN];

static int	in_visual;

static void	dump_scores(void);
static long	env_init(long);
static void	fill_in_blanks(void);
static void	leave(int, char *) __attribute__((__noreturn__));
static void	sigterm(int);
static int	find_driver(void);

/*
 * main:
 *	Main program for local process
 */
int
main(ac, av)
	int	ac;
	char	**av;
{
	int		c;
	extern int	optind;
	extern char	*optarg;
	long		enter_status;
	int		option;
	struct servent	*se;

	enter_status = env_init((long) Q_CLOAK);
	while ((c = getopt(ac, av, "SbcCfh:l:mn:op:qst:w:")) != -1) {
		switch (c) {
		case 'l':	/* rsh compatibility */
		case 'n':
			(void) strlcpy(name, optarg, sizeof name);
			break;
		case 't':
			team = *optarg;
			if (!isdigit(team) && team != ' ') {
				warnx("Team names must be numeric or space");
				team = '-';
			}
			break;
		case 'o':
			Otto_mode = TRUE;
			break;
		case 'm':
			Am_monitor = TRUE;
			break;
		case 'S':
			Show_scores = TRUE;
			break;
		case 'q':	/* query whether hunt is running */
			Query_driver = TRUE;
			break;
		case 'w':
			Send_message = optarg;
			break;
		case 'h':
			Sock_host = optarg;
			break;
		case 'p':
			use_port = optarg;
			Server_port = atoi(use_port);
			break;
		case 'c':
			enter_status = Q_CLOAK;
			break;
		case 'C':
			Use_Color = FALSE;
			break;
		case 'f':
			enter_status = Q_FLY;
			break;
		case 's':
			enter_status = Q_SCAN;
			break;
		case 'b':
			no_beep = !no_beep;
			break;
		default:
		usage:
			fputs(
"usage:\thunt [-bcCfmqsS] [-n name] [-p port] [-t team] [-w message] [[-h] host]\n",
			stderr);
			exit(1);
		}
	}
	if (optind + 1 < ac)
		goto usage;
	else if (optind + 1 == ac)
		Sock_host = av[ac - 1];

	if (Server_port == 0) {
		se = getservbyname(HUNT_PORT_NAME, "udp");
		if (se != NULL)
			Server_port = ntohs(se->s_port);
		else
			Server_port = HUNT_PORT;
	}

	if (Show_scores) {
		dump_scores();
		exit(0);
	}

	if (Query_driver) {
		struct driver		*driver;

		probe_drivers(C_MESSAGE, Sock_host);
		while ((driver = next_driver()) != NULL) {
			printf("%d player%s hunting on %s!\n",
			    driver->response,
			    (driver->response == 1) ? "" : "s",
			    driver_name(driver));
			if (Sock_host)
				break;
		}
		exit(0);
	}
	if (Otto_mode) {
		if (Am_monitor)
			errx(1, "otto mode incompatible with monitor mode");
		(void) strlcpy(name, "otto", sizeof name);
		team = ' ';
	} else
		fill_in_blanks();

	(void) fflush(stdout);
	display_open();

	if (Use_Color)
	    display_init_color();

	in_visual = TRUE;
	if (LINES < SCREEN_HEIGHT || COLS < SCREEN_WIDTH) {
		errno = 0;
		leave(1, "Need a larger window");
	}
	display_clear_the_screen();
	(void) signal(SIGINT, intr);
	(void) signal(SIGTERM, sigterm);
	/* (void) signal(SIGPIPE, SIG_IGN); */

	Daemon_sa_len = 0;
    ask_driver:
	while (!find_driver()) {
		if (Am_monitor) {
			errno = 0;
			leave(1, "No one playing");
		}

		if (Sock_host == NULL) {
			errno = 0;
			leave(1, "huntd not running");
		}

		sleep(3);
	}
	Socket = -1;

	for (;;) {
		if (Socket != -1)
			close(Socket);

		Socket = socket(Daemon.sa_family, SOCK_STREAM, 0);
		if (Socket < 0)
			leave(1, "socket");

		option = 1;
#ifdef SO_USELOOPBACK
		if (setsockopt(Socket, SOL_SOCKET, SO_USELOOPBACK,
		    &option, sizeof option) < 0)
			warn("setsockopt loopback");
#endif

		errno = 0;
		if (connect(Socket, &Daemon, Daemon_sa_len) == -1)  {
			if (errno == ECONNREFUSED)
				goto ask_driver;
			leave(1, "connect");
		}

		do_connect(name, team, enter_status);
		if (Send_message != NULL) {
			do_message(Send_message);
			if (enter_status == Q_MESSAGE)
				break;
			Send_message = NULL;
			continue;
		}
		playit();
		if ((enter_status = quit(enter_status)) == Q_QUIT)
			break;
	}
	leave(0, (char *) NULL);
	/* NOTREACHED */
	return(0);
}

/*
 * Set Daemon to be the address of a hunt driver, or return 0 on failure.
 *
 * We start quietly probing for drivers. As soon as one driver is found
 * we show it in the list. If we run out of drivers and we only have one
 * then we choose it. Otherwise we present a list of the found drivers.
 */
static int
find_driver()
{
	int last_driver, numdrivers, waiting, is_current;
	struct driver *driver;
	int c;
	char buf[80];
	const char *name;

	probe_drivers(Am_monitor ? C_MONITOR : C_PLAYER, Sock_host);

	last_driver = -1;
	numdrivers = 0;
	waiting = 1;
	for (;;) {
		if (numdrivers == 0) {
			/* Silently wait for at least one driver */
			driver = next_driver();
		} else if (!waiting || (driver = 
		    next_driver_fd(STDIN_FILENO)) == (struct driver *)-1) {
			/* We have a key waiting, or no drivers left */
			c = getchar();
			if (c == '\r' || c == '\n' || c == ' ') {
				if (numdrivers == 1)
					c = 'a';
				else if (last_driver != -1)
					c = 'a' + last_driver;
			}
			if (c < 'a' || c >= numdrivers + 'a') {
				display_beep();
				continue;
			}
			driver = &drivers[c - 'a'];
			break;
		}

		if (driver == NULL) {
			waiting = 0;
			if (numdrivers == 0) {
				probe_cleanup();
				return 0;	/* Failure */
			}
			if (numdrivers == 1) {
				driver = &drivers[0];
				break;
			}
			continue;
		}

		/* Use the preferred host straight away. */
		if (Sock_host)
			break;

		if (numdrivers == 0) {
			display_clear_the_screen();
			display_move(1, 0);
			display_put_str("Pick one:");
		}

		/* Mark the last driver we used with an asterisk */
		is_current = (last_driver == -1 && Daemon_sa_len != 0 && 
		    memcmp(&Daemon, &driver->addr, Daemon_sa_len) == 0);
		if (is_current)
			last_driver = numdrivers;

		/* Display it in the list if there is room */
		if (numdrivers < HEIGHT - 3) {
			name = driver_name(driver);
			display_move(3 + numdrivers, 0);
			snprintf(buf, sizeof buf, "%6c %c    %s", 
			    is_current ? '*' : ' ', 'a' + numdrivers, name);
			display_put_str(buf);
		}

		/* Clear the last 'Enter letter' line if any */
		display_move(4 + numdrivers, 0);
		display_clear_eol();

		if (last_driver != -1)
			snprintf(buf, sizeof buf, "Enter letter [%c]: ", 
			    'a' + last_driver);
		else
			snprintf(buf, sizeof buf, "Enter letter: ");

		display_move(5 + numdrivers, 0);
		display_put_str(buf);
		display_refresh();

		numdrivers++;
	}

	display_clear_the_screen();
	Daemon = driver->addr;
	Daemon_sa_len = SA_LEN(&Daemon);

	probe_cleanup();
	return 1;		/* Success */
}

static void
dump_scores()
{
	struct	driver *driver;
	int	s, cnt, i;
	char	buf[1024];

	probe_drivers(C_SCORES, Sock_host);
	while ((driver = next_driver()) != NULL) {
		printf("\n%s:\n", driver_name(driver));
		fflush(stdout);

		if ((s = socket(driver->addr.sa_family, SOCK_STREAM, 0)) < 0) {
			warn("socket");
			continue;
		}
		if (connect(s, &driver->addr, SA_LEN(&(driver->addr))) < 0) {
			warn("connect");
			close(s);
			continue;
		}
		while ((cnt = read(s, buf, sizeof buf)) > 0) {
			/* Whittle out bad characters */
			for (i = 0; i < cnt; i++)
				if ((buf[i] < ' ' || buf[i] > '~') &&
				    buf[i] != '\n' && buf[i] != '\t')
					buf[i] = '?';
			fwrite(buf, cnt, 1, stdout);
		}
		if (cnt < 0)
			warn("read");
		(void)close(s);
		if (Sock_host)
			break;
	}
	probe_cleanup();
}


/*
 * bad_con:
 *	We had a bad connection.  For the moment we assume that this
 *	means the game is full.
 */
void
bad_con()
{
	leave(1, "lost connection to huntd");
}

/*
 * bad_ver:
 *	version number mismatch.
 */
void
bad_ver()
{
	errno = 0;
	leave(1, "Version number mismatch. No go.");
}

/*
 * sigterm:
 *	Handle a terminate signal
 */
static void
sigterm(dummy)
	int dummy;
{
	leave(0, (char *) NULL);
}

/*
 * rmnl:
 *	Remove a '\n' at the end of a string if there is one
 */
static void
rmnl(s)
	char	*s;
{
	char	*cp;

	cp = strrchr(s, '\n');
	if (cp != NULL)
		*cp = '\0';
}

/*
 * intr:
 *	Handle a interrupt signal
 */
void
intr(dummy)
	int dummy;
{
	int	ch;
	int	explained;
	int	y, x;

	(void) signal(SIGINT, SIG_IGN);
	display_getyx(&y, &x);
	display_move(HEIGHT, 0);
	display_put_str("Really quit? ");
	display_clear_eol();
	display_refresh();
	explained = FALSE;
	for (;;) {
		ch = getchar();
		if (isupper(ch))
			ch = tolower(ch);
		if (ch == 'y') {
			if (Socket != 0) {
				(void) write(Socket, "q", 1);
				(void) close(Socket);
			}
			leave(0, (char *) NULL);
		}
		else if (ch == 'n') {
			(void) signal(SIGINT, intr);
			display_move(y, x);
			display_refresh();
			return;
		}
		if (!explained) {
			display_put_str("(Yes or No) ");
			display_refresh();
			explained = TRUE;
		}
		display_beep();
		display_refresh();
	}
}

/*
 * leave:
 *	Leave the game somewhat gracefully, restoring all current
 *	tty stats.
 */
static void
leave(eval, mesg)
	int	eval;
	char	*mesg;
{
	int saved_errno;

	saved_errno = errno;
	if (in_visual) {
		display_move(HEIGHT, 0);
		display_refresh();
		display_end();
	}
	errno = saved_errno;

	if (errno == 0 && mesg != NULL)
		errx(eval, mesg);
	else if (mesg != NULL)
		err(eval, mesg);
	exit(eval);
}

/*
 * env_init:
 *	initialise game parameters from the HUNT envvar
 */
static long
env_init(enter_status)
	long	enter_status;
{
	int	i;
	char	*envp, *envname, *s;

	/* Map all keys to themselves: */
	for (i = 0; i < 256; i++)
		map_key[i] = (char) i;

	envname = NULL;
	if ((envp = getenv("HUNT")) != NULL) {
		while ((s = strpbrk(envp, "=,")) != NULL) {
			if (strncmp(envp, "cloak,", s - envp + 1) == 0) {
				enter_status = Q_CLOAK;
				envp = s + 1;
			}
			else if (strncmp(envp, "scan,", s - envp + 1) == 0) {
				enter_status = Q_SCAN;
				envp = s + 1;
			}
			else if (strncmp(envp, "fly,", s - envp + 1) == 0) {
				enter_status = Q_FLY;
				envp = s + 1;
			}
			else if (strncmp(envp, "nobeep,", s - envp + 1) == 0) {
				no_beep = TRUE;
				envp = s + 1;
			}
			else if (strncmp(envp, "name=", s - envp + 1) == 0) {
				envname = s + 1;
				if ((s = strchr(envp, ',')) == NULL) {
					*envp = '\0';
					strlcpy(name, envname, sizeof name);
					break;
				}
				*s = '\0';
				strlcpy(name, envname, sizeof name);
				envp = s + 1;
			}
			else if (strncmp(envp, "port=", s - envp + 1) == 0) {
				use_port = s + 1;
				Server_port = atoi(use_port);
				if ((s = strchr(envp, ',')) == NULL) {
					*envp = '\0';
					break;
				}
				*s = '\0';
				envp = s + 1;
			}
			else if (strncmp(envp, "host=", s - envp + 1) == 0) {
				Sock_host = s + 1;
				if ((s = strchr(envp, ',')) == NULL) {
					*envp = '\0';
					break;
				}
				*s = '\0';
				envp = s + 1;
			}
			else if (strncmp(envp, "message=", s - envp + 1) == 0) {
				Send_message = s + 1;
				if ((s = strchr(envp, ',')) == NULL) {
					*envp = '\0';
					break;
				}
				*s = '\0';
				envp = s + 1;
			}
			else if (strncmp(envp, "team=", s - envp + 1) == 0) {
				team = *(s + 1);
				if (!isdigit(team))
					team = ' ';
				if ((s = strchr(envp, ',')) == NULL) {
					*envp = '\0';
					break;
				}
				*s = '\0';
				envp = s + 1;
			}			/* must be last option */
			else if (strncmp(envp, "mapkey=", s - envp + 1) == 0) {
				for (s = s + 1; *s != '\0'; s += 2) {
					map_key[(unsigned int) *s] = *(s + 1);
					if (*(s + 1) == '\0') {
						break;
					}
				}
				*envp = '\0';
				break;
			} else {
				*s = '\0';
				printf("unknown option %s\n", envp);
				if ((s = strchr(envp, ',')) == NULL) {
					*envp = '\0';
					break;
				}
				envp = s + 1;
			}
		}
		if (*envp != '\0') {
			if (envname == NULL)
				strlcpy(name, envp, sizeof name);
			else
				printf("unknown option %s\n", envp);
		}
	}
	return enter_status;
}

/*
 * fill_in_blanks:
 *	quiz the user for the information they didn't provide earlier
 */
static void
fill_in_blanks()
{
	int	i;
	char	*cp;

again:
	if (name[0] != '\0') {
		printf("Entering as '%s'", name);
		if (team != ' ' && team != '-')
			printf(" on team %c.\n", team);
		else
			putchar('\n');
	} else {
		printf("Enter your code name: ");
		if (fgets(name, sizeof name, stdin) == NULL)
			exit(1);
	}
	rmnl(name);
	if (name[0] == '\0') {
		printf("You have to have a code name!\n");
		goto again;
	}
	for (cp = name; *cp != '\0'; cp++)
		if (!isprint(*cp)) {
			name[0] = '\0';
			printf("Illegal character in your code name.\n");
			goto again;
		}
	if (team == '-') {
		printf("Enter your team (0-9 or nothing): ");
		i = getchar();
		if (isdigit(i))
			team = i;
		else if (i == '\n' || i == EOF || i == ' ')
			team = ' ';
		/* ignore trailing chars */
		while (i != '\n' && i != EOF)
			i = getchar();
		if (team == '-') {
			printf("Teams must be numeric.\n");
			goto again;
		}
	}
}
